[Sema] Suggest missing format attributes

vvuksanovic opened the pull request:
Implements the `-Wmissing-format-attribute` diagnostic as a subgroup of `-Wformat-nonliteral`. It suggests adding format attributes to function declarations that call other format functions and pass the format string to them.

This is an updated implementation of #105479.

vvuksanovic committed:
[Sema] Suggest missing format attributes

Implements the `-Wmissing-format-attribute` diagnostic. It suggests
adding format attributes to function declarations that call other
format functions and pass the format string and arguments to them.

Co-authored-by: Budimir Arandjelovic <budimir.arandjelovic@htecgroup.com>

Fznamznon commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -401,6 +401,8 @@ Improvements to Clang's diagnostics
 - Clang now emits a diagnostic in case `vector_size` or `ext_vector_type`
   attributes are used with a negative size (#GH165463).
+- Clang now detects potential missing format attributes on function declarations
+  when calling format functions. (#GH60718)
```
https://github.com/llvm/llvm-project/issues/60718 seems to be closed already?

Sirraide commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3478,6 +3478,11 @@ def err_format_attribute_result_not : Error<"function does not return %0">;
+    : Warning<"diagnostic behavior may be improved by adding the '%0' format "
+              "attribute to the declaration of %1">,
+      InGroup<DiagGroup<"missing-format-attribute">>,
+      DefaultIgnore;
```
Please change the formatting of this to be more similar to the other diagnostics (clang-format doesn’t do this properly unfortunately...)

vvuksanovic replied:
Should be okay now. clang-format always tries to reformat it.

Sirraide reviewed (commented):
One concern I have with this is that `__attribute__(())` is non-standard syntax, so suggesting that in a fix-it is... not great for portability, because e.g. MSVC doesn’t support that syntax at all. We probably want to only emit this diagnostic (or perform the check at all really) if we’re in `-std=gnuXY`/`-std=gnu++XY` mode.

Sirraide commented:
> We probably want to only emit this diagnostic (or perform the check at all really) if we’re in `-std=gnuXY`/`-std=gnu++XY` mode.

Also, in non-GNU C23/C++11 mode, we can just suggest `[[gnu::format]]` instead

vvuksanovic committed:
Switch to portable attribute syntax

vvuksanovic commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -6635,10 +6636,11 @@ static StringLiteralCheckType checkFormatStringExpr(
+          return checkFormatStringExpr(S, ReferenceFormatString, Init, Args,
+                                       APK, format_idx, firstDataArg, Type,
+                                       CallType,
+                                       /*InFunctionCall*/ false, CheckedVarArgs,
```
Done

apple-fcloutier commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -7001,6 +7005,85 @@ bool Sema::CheckFormatString(const FormatMatchesAttr *Format,
+                                         SourceLocation Loc) {
+  const FunctionDecl *Caller = S->getCurFunctionDecl();
+  if (!Caller)
+    return;
```
This will always fail in Objective-C because ObjCMethodDecl does not inherit from FunctionDecl. Would you add an Objective-C test to ensure this works? If you're not super familiar with it, the declaration/use syntax should be:

```
#include <stdarg.h>

@interface Foo
-(void)myprintf:(const char *)fmt, ... __attribute__((format(printf, 1, 2)));
-(void)myvprintf:(const char *)fmt list:(va_list)ap __attribute__((format(printf, 1, 0)));
@end

void bar(Foo *f) {
    [f myprintf:"hello %i", 123];
}

void baz(Foo *f, const char *fmt, va_list ap) {
    [f myvprintf:fmt list:ap];
}
```

Happy to answer other questions you may have.

vvuksanovic replied:
I'm not familiar with objective-c and there doesn't seem to be much documentation regarding attributes, but I think I have a working implementation. Just one question: is there an implicit 'this' argument on obj-c methods?

I'll have to look into blocks to see how they fit in here.

apple-fcloutier replied:
Objective-C methods have two implicit arguments (`self` and `_cmd`), but IIRC index 1 corresponds to the first explicit argument (ie, _not_ like how it's done for C++).

vvuksanovic replied:
Added support for Objective-C. Let me know if there are other cases I should test. Also, I wasn't sure which attribute syntax is supported, so I used the GNU extension.

apple-fcloutier replied:
Seems to cover it, thank you!

apple-fcloutier commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3478,6 +3478,10 @@ def err_format_attribute_result_not : Error<"function does not return %0">;
+def warn_missing_format_attribute : Warning<
+  "diagnostic behavior may be improved by adding the '%0' format "
+  "attribute to the declaration of %1">,
+  InGroup<DiagGroup<"missing-format-attribute">>, DefaultIgnore;
```
Can you help me put in context how -Wmissing-format-attribute works with GCC and its relationship with -Wformat, -Wformat=2 and -Wformat-nonliteral? It's kind of a least favorite way forward to add new off-by-default diagnostics.

Here's my thinking (before knowing how GCC deals with this):

* Could -Wmissing-format-attribute be enabled by default and/or be part of -Wformat? (Do we need an adult to confirm the bar for adding on-by-default diagnostics? I think this should be OK given that this shouldn't have false positives at all.)
* If -Wmissing-format-attribute cannot be enabled by default, what are our options for splitting the difference between it and -Wformat-nonliteral? Should -Wmissing-format-attribute be part of -Wformat-nonliteral so that users of -Wformat-nonliteral get it when possible?

apple-fcloutier reviewed (changes requested):
Thanks for addressing the comments I had on the last iteration. I would mostly like us to resolve quality-of-life issues with -Wmissing-format-attribute and how it combines with other format warnings to proceed.

Sirraide commented:
> Another possible issue is that clang supports the format attribute on non-variadic functions, which GCC doesn't. This diagnostic can suggest adding the attribute to such functions, which would cause an error when compiling with GCC. Should I just disable that case?

That’s a good point; we probably shouldn’t issue a fixit that when applied causes GCC to error... we might still be able to add an implicit format attribute as suggested by https://github.com/llvm/llvm-project/pull/166738#discussion_r2504843326 though.

apple-fcloutier commented:
As far as I'm concerned, `-Wmissing-format-attribute` is the subset of `-Wformat-nonliteral` that we can fix it by adding a format attribute. This change is valuable and I want it to land, but I think that we need to close the gaps with `-Wformat-nonliteral` to make sure we're consistent. This mostly means ensuring you don't get both a `-Wmissing-format-attribute` and a `-Wformat-nonliteral` diagnostic for the same code. And because of that relationship, I think we should also make `-Wmissing-format-attribute` a subgroup of `-Wformat-nonliteral`, as it will subsume diagnostics that -Wformat-nonliteral previously emitted.

This does mean code that was previously clean with `-Wformat-nonliteral` might now get diagnostics for strftime (since enabling `-Wformat-nonliteral` would enable `-Wmissing-format-attribute` and `-Wformat-nonliteral` did not trip for strftime), but I think that's fine. It also means that users could do `-Wformat-nonliteral -Wno-missing-format-attribute` and get fewer diagnostics than before, but there's a lot of silly diagnostic combinations and I don't think that should stop us.

(Or, alternatively, change my mind!)

erichkeane reviewed (commented):
I'm happy when @apple-fcloutier is here, I don't intend to review this again unless separately pinged, consider my feedback to be theirs/vice-versa.

vvuksanovic committed:
Address review comments

- Check now works or Objective-C
- Add attribute arguments to diagnostic message
- Rename CallerParamIdx to CallerFormatParamIdx
- Fix typo

vvuksanovic committed:
Emit fixit only when supported

vvuksanovic commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -7001,6 +7005,85 @@ bool Sema::CheckFormatString(const FormatMatchesAttr *Format,
+                                     AttrSuffix)
+                                        .str());
+  S->Diag(Caller->getFirstDecl()->getLocation(), diag::note_callee_decl)
+      << Caller;
```
I must have searched by the diagnostic message. Changed to `note_entity_declared_at`. Or is it preferred to create a note specific to this diagnostic?

vvuksanovic commented:
> As far as I'm concerned, `-Wmissing-format-attribute` is the subset of `-Wformat-nonliteral` that we can fix it by adding a format attribute. This change is valuable and I want it to land, but I think that we need to close the gaps with `-Wformat-nonliteral` to make sure we're consistent. This mostly means ensuring you don't get both a `-Wmissing-format-attribute` and a `-Wformat-nonliteral` diagnostic for the same code. And because of that relationship, I think we should also make `-Wmissing-format-attribute` a subgroup of `-Wformat-nonliteral`, as it will subsume diagnostics that -Wformat-nonliteral previously emitted.

Let me make sure I understand. The idea is to emit `-Wmissing-format-attribute` instead of `-Wformat-nonliteral` when possible, otherwise `-Wformat-nonliteral` would trigger. If so, I agree with that.

apple-fcloutier commented:
> Let me make sure I understand. The idea is to emit `-Wmissing-format-attribute` instead of `-Wformat-nonliteral` when possible, otherwise `-Wformat-nonliteral` would trigger. If so, I agree with that.

Yes, exactly (in addition to making -Wmissing-format-attribute a subgroup of -Wformat-nonliteral so that users of -Wformat-nonliteral get it too).

vvuksanovic committed:
Fix ObjC fixit location

vvuksanovic committed:
Suggest format_matches when applicable

vvuksanovic committed:
Add implicit attribute

vvuksanovic committed:
Make diagnostic a subgroup of format-nonliteral

apple-fcloutier reviewed (changes requested):
Thanks for the format_matches support (and reviewing again, I now see that ReleaseNotes.rst was updated very early on). The last bits that I'm still tracking:

1. -Wmissing-format-attribute should be a subgroup of -Wformat-nonliteral (should have a test that with both enabled, you only get -Wmissing-format-attribute).
2. We should not add the implicit attribute if diag::warn_missing_format_attribute is ignored, since it would create downstream diagnostics that cannot be disabled in any reasonable way. We should have a test that with it enabled, you do get diagnostics caused by the implicit attribute, and with it disabled, you do not.

I expect these are the last changes I'll request, everything else looks good.

----

Example for point 1:

```c
// clang -fsyntax-only -Wformat-nonliteral -Wmissing-format-attribute test.c
#include <stdarg.h>

__attribute__((format(printf, 1, 0)))
int vprintf(const char *, va_list);

void foo(const char *fmt, ...) {
	va_list ap;
	va_start(ap, fmt);
	vprintf(fmt, ap);
	// warning: diagnostic behavior may be improved by adding the 'format(printf, 1, 2)' attribute to the declaration of 'foo' [-Wmissing-format-attribute]
	// warning: format string is not a string literal [-Wformat-nonliteral]
}
```

The expected result is that we have just the -Wmissing-format-attribute since they are both complaining about the same thing but -Wmissing-format-attribute knows how to fix it. Enabling -Wformat-nonliteral should enable -Wmissing-format-attribute as well since it is now the better diagnostic for that situation.

---

Example for point 2:

```c
// clang -fsyntax-only -Wformat
#include <stdarg.h>

__attribute__((format(printf, 1, 0)))
int vprintf(const char *, va_list);

void foo(const char *fmt, ...) {
	va_list ap;
	va_start(ap, fmt);
	vprintf(fmt, ap);
}

void bar(void) {
	foo("%g\n", 123);
	// warning: format specifies type 'double' but the argument has type 'int' [-Wformat]
}
```

`foo` is not attributed in source and there is no clear explanation for what's going on; if there's ever a false positive, there is no way to prevent clang from adding the format attribute implicitly to `foo`. The expected result is that if -Wmissing-format-attribute is disabled, then we should not implicitly add the format attribute.

vvuksanovic committed:
Cleanup

vvuksanovic committed:
Merge remote-tracking branch 'upstream/main' into missing-format-attribute

vvuksanovic committed:
Fix cpp test

vvuksanovic committed:
Fix objc test

apple-fcloutier reviewed (approved):
Thanks, I think we have everything down now.

> I am not a fan of the diagnostic message for blocks, so let me know if you have a suggestion.

Format attributes on blocks are messy because they're declaration attributes rather than type attributes. I don't think that we need to spend more time on it than you already have.

I have 3 last cleanup-grade requests. I'm confident I won't need to review the changes again, so I'm approving early:

* Including "clang/Basic/CharInfo.h" is no longer necessary (I think).
* We need to pass a real source location to `isIgnored(..., SourceLocation())` to support diagnostic control through pragmas.
* We should change the test file names from attr-format-* to format-attr-* so that they are alphabetically grouped with the other format attribute tests.

vvuksanovic committed:
Cleanup

vvuksanovic commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -7093,9 +7105,143 @@ bool Sema::CheckFormatString(const FormatMatchesAttr *Format,
+    unsigned FirstDataArg, FormatStringType FormatType, unsigned CallerParamIdx,
+    SourceLocation Loc) {
+  if (S->getDiagnostics().isIgnored(diag::warn_missing_format_attribute,
+                                    SourceLocation()))
```
I changed that by accident, it should be `Loc`. Fixed now,

vvuksanovic committed:
Fix formatting

vvuksanovic commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -40,6 +41,7 @@
 #include "clang/AST/TypeLoc.h"
 #include "clang/AST/UnresolvedSet.h"
 #include "clang/Basic/AddressSpaces.h"
+#include "clang/Basic/CharInfo.h"
```
Correct. Removed the include.

vvuksanovic commented on clang/test/Sema/format-attr-missing.m:
```diff
@@ -0,0 +1,68 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -Wmissing-format-attribute %s
```
Done.

vvuksanovic commented:
Okay then, thank you for the review and the help.

If nobody else has any comments, can you or someone else merge the PR? I don't have commit access.

apple-fcloutier commented:
@vvuksanovic, did you have a person in mind to merge the PR when you opened it? I gave up my commit privileges when there was an overhaul because my contributions were too sparse.

AaronBallman commented:
> @vvuksanovic, did you have a person in mind to merge the PR when you opened it? I gave up my commit privileges when there was an overhaul because my contributions were too sparse.

I'll get it, I didn't realize you didn't have commit privs any longer. :-)
